### PR TITLE
feat: [IA-268] Support infinite scroll

### PR DESCRIPTION
**Pull request by pietro909**
## Short description
This PR finalizes the pagination logic by handling errors, end of list, loading state(s).
I moved the MessageDetails to its own folder under `paginated/` where is now in a WIP state.
Many other things where descoped due to the sheer amount of files touched and dependencies with the backend.

### Questions you may have

* why pull-to-refresh doesn't load the previous page? See [IA-269](https://pagopa.atlassian.net/browse/IA-269)
* why message details view is empty? See [IA-465](https://pagopa.atlassian.net/browse/IA-465)
* new _paid_ status doesn't work? See [IA-417](https://pagopa.atlassian.net/browse/IA-417)

### Under feature flag
- Tabs [are gone](ts/components/messages/HomeTabs.tsx) for good
- the [MessageList component](ts/components/messages/paginated/MessageList/index.tsx) is acting as the main hub for message management, taking care of error handling, filtering, pagination
- three new (tested) sagas with dedicated actions are available for managing the [reload](ts/sagas/messages/watchReloadAllMessages.ts), [next](ts/sagas/messages/watchLoadNextPageMessages.ts), [previous](ts/sagas/messages/watchLoadPreviousPageMessages.ts)

### To Production code
- extracted the logic for response management to [separate util](https://github.com/pagopa/io-app/pull/3501/files?authenticity_token=uU9B%2BtiU4wHpIP0XkqrU9UcF46mEhvxCR6VGhWiJ%2BGLSRewU7jLH3421CwKbC11FfuOIvCYcJ4trubFGenwtlA%3D%3D&file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx&hide-deleted-files=true#diff-d7c2f242fcdf72bb57662d9fd0a26b1b5ad62333e449256fe512c1070f8f3ed5R25) properly tested
- former [loadMessage](https://github.com/pagopa/io-app/pull/3501/files?authenticity_token=uU9B%2BtiU4wHpIP0XkqrU9UcF46mEhvxCR6VGhWiJ%2BGLSRewU7jLH3421CwKbC11FfuOIvCYcJ4trubFGenwtlA%3D%3D&file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx&hide-deleted-files=true#diff-5b727d84caebb631cf6843fbd8f3595a97801341cee749b8d9535aabdbc12460R34) is deprecated to better separate the new type for message details (see the next paragraph)
- added many, **many**, new unit tests

## How to test

Run the app with the flag on and configure your server to return an error once in a while. In the video below, I've configured it to reject 50% of the requests to `/messages`. That's why it shows that often.
Upon error, you are able to scroll again to retry the same action.

https://user-images.githubusercontent.com/2094604/142208914-cb864339-1428-4729-807f-dffaa7f0b090.mov

### CODE REVIEW

**Review by Undermaken — CHANGES_REQUESTED** [review]
@pietro909 what do you think to increase the page size to 12 or a number that is related with the device height?

During my tests I noticed that if the number of items can be included (with some margins) in the screen height, the list end listener is invoked and the app load 2 pages instead of one

In the other hand, if we increase the page size, it doesn't happen

Here a video proof of that

https://user-images.githubusercontent.com/822471/142662505-c4d6c965-6364-4b20-80c5-5def94bf5047.mp4

File: ts/screens/messages/paginated/MessagesHomeScreen.tsx [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    // TODO: https://pagopa.atlassian.net/browse/IA-463
+    // if message is a GreenPass, navigate to the dedicated navigator
+    //   navigateToEuCovidCertificate(
+    //     message.content.eu_covid_cert.auth_code as EUCovidCertificateAuthCode,
+    //     message.id
+    //   );
+    navigation.dispatch(
+      navigateToPaginatedMessageDetailScreenAction({ message })
+    );
+  };
```
[the remark this answers is not in the document]
Comment (pietro909) [reply ts/screens/messages/paginated/MessagesHomeScreen.tsx:69]: I actually got the code for this from [our docs](https://docs.google.com/document/d/1SD02SfpnfEnugFl3-BeX8nmCv-7PpYrq48iRi4-xH_M/edit#heading=h.r9i3khvliykk) written by @fabriziofff 
Here I'm using what is supposed to be our _future_ way of doing things.

[the remark this answers is not in the document]
Comment (fabriziofff) [reply ts/screens/messages/paginated/MessagesHomeScreen.tsx:69]: If we want execute a navigation action already defined, they are both the same ([`navigate`](https://reactnavigation.org/docs/4.x/navigation-prop#navigate---link-to-other-screens) uses [`dispatch`](https://reactnavigation.org/docs/4.x/navigation-prop#dispatch---send-an-action-to-the-router)). 
Moreover `dispatch` allows to dispatch even more complex actions such as `StackActions.popToTop`.
I think it's a matter of preference, I prefer dispatch due to its generic nature.

[the remark this answers is not in the document]
Comment (Undermaken) [reply ts/screens/messages/paginated/MessagesHomeScreen.tsx:69]: Thank you! I thought I did it right in another PR. Maybe I have to change my code following this one 😅

File: ts/components/messages/paginated/MessageList/index.tsx [hunk 2]
```diff
@@ -198,26 +229,27 @@ const MessageList = ({
...
-        }
+        ListEmptyComponent={() => {
+          if (error !== undefined) {
+            return <ErrorLoadingComponent />;
+          }
+          if (ListEmptyComponent) {
+            return <ListEmptyComponent />;
+          }
+          return null;
+        }}
```
[the remark this answers is not in the document]
Comment (pietro909) [reply ts/components/messages/paginated/MessageList/index.tsx]: Makes sense. I've extracted a helper method for now: https://github.com/pagopa/io-app/pull/3510/commits/c2631266ce1f795bac0335f1ae21388c9ee7272a#diff-5d2069c099684d6a56a86df750b25c4cbf098b242c751215afab02b8f5a592a6R64

File: ts/components/messages/paginated/MessageList/index.tsx [hunk 3]
```diff
@@ -142,11 +157,15 @@ const MessageList = ({
 
   const [isFirstLoad, setIsFirstLoad] = useState(isIos);
 
-  const onEndReached = () => {
-    if (nextCursor && !isLoadingMore) {
-      loadNextPage(nextCursor);
+  useEffect(() => {
+    if (error) {
+      Toast.show({
```
[the remark this answers is not in the document]
Comment (pietro909) [reply ts/components/messages/paginated/MessageList/index.tsx]: Brilliant! It's also easier to test https://github.com/pagopa/io-app/pull/3510/commits/c2631266ce1f795bac0335f1ae21388c9ee7272a#diff-a0ab177007bd8f5d749d9e7494df01aaddde642f3d379c200e99e9c6048b9b17R159

File: ts/components/messages/paginated/MessageList/index.tsx [hunk 4]
```diff
@@ -198,26 +229,27 @@ const MessageList = ({
...
+          if (error !== undefined) {
+            return <ErrorLoadingComponent />;
+          }
+          if (ListEmptyComponent) {
+            return <ListEmptyComponent />;
+          }
+          return null;
+        }}
         data={messages}
         initialNumToRender={10}
```
[the remark this answers is not in the document]
Comment (pietro909) [reply ts/components/messages/paginated/MessageList/index.tsx]: Good point. [Fixed](https://github.com/pagopa/io-app/pull/3510/commits/c2631266ce1f795bac0335f1ae21388c9ee7272a#diff-a0ab177007bd8f5d749d9e7494df01aaddde642f3d379c200e99e9c6048b9b17R234).

File: ts/screens/messages/paginated/MessagesHomeScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+        searchText
+          .map(_ =>
+            _.length < MIN_CHARACTER_SEARCH_TEXT ? (
+              <SearchNoResultMessage errorType="InvalidSearchBarText" />
+            ) : (
+              <MessagesSearch
+                messages={allMessages}
+                searchText={_}
+                renderSearchResults={results => (
+                  <MessageList filteredMessages={results} />
```
[the remark this answers is not in the document]
Comment (pietro909) [reply ts/screens/messages/paginated/MessagesHomeScreen.tsx]: It is indeed :-) - Done in https://github.com/pagopa/io-app/pull/3510/commits/c2631266ce1f795bac0335f1ae21388c9ee7272a#diff-0f5e4629af7b77f0b21af3a54fe1d4d21314eb44c8a54193316d9d80ba2654bfR38

File: ts/sagas/messages/watchReloadAllMessages.ts [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+          reloadAllMessagesAction.success({
+            messages: items.map(toUIMessage),
+            pagination: { previous: prev, next }
+          }),
+        error => reloadAllMessagesAction.failure(error)
+      );
+
+      yield put(nextAction);
+    } catch (error) {
+      yield put(reloadAllMessagesAction.failure(error));
```
[the remark this answers is not in the document]
Comment (pietro909) [reply ts/sagas/messages/watchReloadAllMessages.ts]: Good catch! I've updated this and the other watchers https://github.com/pagopa/io-app/pull/3510/commits/c2631266ce1f795bac0335f1ae21388c9ee7272a#diff-0f5e4629af7b77f0b21af3a54fe1d4d21314eb44c8a54193316d9d80ba2654bfR38

File: ts/sagas/messages/watchLoadPreviousPageMessages.ts [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+      );
+
+      yield put(nextAction);
+    } catch (error) {
+      yield put(loadPreviousPageMessagesAction.failure(error));
+    }
+  };
+}
+
+export const testTryLoadPreviousPageMessages = isDevEnv
```
[the remark this answers is not in the document]
Comment (pietro909) [reply ts/sagas/messages/watchLoadPreviousPageMessages.ts]: Totally :-) I've updated also the other watchers.

File: ts/screens/messages/paginated/MessagesHomeScreen.tsx [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    reloadFirstPage();
+    const listener = navigation.addListener("didFocus", () => {
+      setStatusBarColorAndBackground(
+        "dark-content",
+        customVariables.colorWhite
+      );
+    });
+    return () => {
+      listener.remove();
+    };
```
[the remark this answers is not in the document]
Comment (pietro909) [reply ts/screens/messages/paginated/MessagesHomeScreen.tsx:81]: Tried and it doesn't work as expected: `useActionOnFocus` calls the action only at the very first render, while the listener on `didFocus` calls it for every time the tab is selected.
We need it to run every time it is focused to update the status bar.

See it in action:


https://user-images.githubusercontent.com/2094604/142868645-21b38837-4ca7-42ed-b5f2-f93d00b65045.mov

**Comment by pietro909** [comment]
> what do you think to increase the page size to 12 or a number that is related with the device height?
https://github.com/pagopa/io-app/pull/3510/commits/c2631266ce1f795bac0335f1ae21388c9ee7272a#diff-0f5e4629af7b77f0b21af3a54fe1d4d21314eb44c8a54193316d9d80ba2654bfR38

**Review by Undermaken — APPROVED** [review]

**Pull request merged**